### Import Packages

In [1]:
import pandas as pd
import numpy as np
import datetime
import matplotlib.pyplot as plt
%matplotlib inline

from weather import forecast

## Seattle Collision Data

https://data-seattlecitygis.opendata.arcgis.com/datasets/collisions/data?geometry=-122.526%2C47.676%2C-122.198%2C47.717&page=4

01/01/2014 - 03/29/2018

data with weather info available 01/01/2014 - 01/30/2018

(lat, lon) values have 6 decimal places (precision at individual human level)

### read in the original dataset

In [3]:
df_collision = pd.read_csv('../data/Collisions.csv')

In [4]:
df_collision.shape

(196116, 40)

#### check one example

In [5]:
df_collision.head(1).T

0
X                                                         -122.342
Y                                                          47.6276
objectid                                                 162244138
inckey                                                      233006
coldetkey                                                   233406
addrtype                                                     Block
collisiontype                                              Head On
exceptrsncode                                                     
exceptrsndesc                                                  NaN
fatalities                                                       0
inattentionind                                                 NaN
incdate                                   2016-11-04T00:00:00.000Z
incdttm                                       11/4/2016 9:55:00 AM
injuries                                                         0
intkey                                                         NaN
junctiontype               Mid-Block (not related to intersection)
lightcond                                                 Daylight
location                 DEXTER AVE N BETWEEN ALOHA ST AND WARD ST
pedcount                                                         0
pedcylcount                                                      0
pedrownotgrnt                                                  NaN
personcount                                                      2
reportno                                                   E605707
roadcond                                                       Dry
sdot_colcode                                                    11
sdot_coldesc     MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...
sdotcolnum                                                     NaN
seriousinjuries                                                  0
severitycode                                                     1
severitydesc                        Property Damage Only Collision
speeding                                                       NaN
st_colcode                                                      30
st_coldesc                    From Opposite Direction - All Others
status                                                     Matched
underinfl                                                        N
vehcount                                                         2
weather                                     Clear or Partly Cloudy
seglanekey                                                       0
crosswalkkey                                                     0
hitparkedcar                                                     N

#### check column names

In [6]:
df_collision.columns.values

array(['X', 'Y', 'objectid', 'inckey', 'coldetkey', 'addrtype',
       'collisiontype', 'exceptrsncode', 'exceptrsndesc', 'fatalities',
       'inattentionind', 'incdate', 'incdttm', 'injuries', 'intkey',
       'junctiontype', 'lightcond', 'location', 'pedcount', 'pedcylcount',
       'pedrownotgrnt', 'personcount', 'reportno', 'roadcond',
       'sdot_colcode', 'sdot_coldesc', 'sdotcolnum', 'seriousinjuries',
       'severitycode', 'severitydesc', 'speeding', 'st_colcode',
       'st_coldesc', 'status', 'underinfl', 'vehcount', 'weather',
       'seglanekey', 'crosswalkkey', 'hitparkedcar'], dtype=object)

In [7]:
columns = ['X', 'Y', 'addrtype', 'incdate', 'incdttm', 'lightcond', 'location', 'sdot_coldesc'
          , 'severitydesc', 'speeding', 'weather']

df_collision.loc[:, columns].head()

X          Y      addrtype                   incdate  \
0 -122.342323  47.627572         Block  2016-11-04T00:00:00.000Z   
1 -122.301639  47.630771         Block  2016-12-01T00:00:00.000Z   
2 -122.290784  47.719253  Intersection  2016-12-23T00:00:00.000Z   
3 -122.322535  47.623135         Block  2016-12-21T00:00:00.000Z   
4 -122.376803  47.689665         Block  2016-12-23T00:00:00.000Z   

                  incdttm lightcond  \
0    11/4/2016 9:55:00 AM  Daylight   
1   12/1/2016 10:33:00 AM  Daylight   
2  12/23/2016 12:48:00 PM  Daylight   
3   12/21/2016 4:13:00 PM  Daylight   
4   12/23/2016 6:33:00 AM      Dawn   

                                            location  \
0          DEXTER AVE N BETWEEN ALOHA ST AND WARD ST   
1      24TH AVE E BETWEEN E HIGHLAND DR AND E LEE ST   
2                        35TH AVE NE AND NE 125TH ST   
3  E REPUBLICAN ST BETWEEN BOYLSTON AVE E AND HAR...   
4      15TH AVE NW BETWEEN NW 83RD ST AND NW 85TH ST   

                                        sdot_coldesc  \
0  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
1  MOTOR VEHICLE STRUCK MOTOR VEHICLE, LEFT SIDE ...   
2  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
3  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
4  MOTOR VEHICLE STRUCK MOTOR VEHICLE, LEFT SIDE ...   

                     severitydesc speeding                 weather  
0  Property Damage Only Collision      NaN  Clear or Partly Cloudy  
1  Property Damage Only Collision      NaN                Overcast  
2  Property Damage Only Collision      NaN                 Snowing  
3  Property Damage Only Collision      NaN  Clear or Partly Cloudy  
4  Property Damage Only Collision      NaN                 Raining

In [8]:
df_collision.loc[:, columns].head(1).T

0
X                                                      -122.342
Y                                                       47.6276
addrtype                                                  Block
incdate                                2016-11-04T00:00:00.000Z
incdttm                                    11/4/2016 9:55:00 AM
lightcond                                              Daylight
location              DEXTER AVE N BETWEEN ALOHA ST AND WARD ST
sdot_coldesc  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...
severitydesc                     Property Damage Only Collision
speeding                                                    NaN
weather                                  Clear or Partly Cloudy

In [9]:
columns = ['X', 'Y', 'incdttm', 'weather']
df_collision.loc[:, columns].head(5)

X          Y                 incdttm                 weather
0 -122.342323  47.627572    11/4/2016 9:55:00 AM  Clear or Partly Cloudy
1 -122.301639  47.630771   12/1/2016 10:33:00 AM                Overcast
2 -122.290784  47.719253  12/23/2016 12:48:00 PM                 Snowing
3 -122.322535  47.623135   12/21/2016 4:13:00 PM  Clear or Partly Cloudy
4 -122.376803  47.689665   12/23/2016 6:33:00 AM                 Raining

### Check Missing Values in the Dataset

In [10]:
df_collision.speeding.isna().value_counts()

True     187386
False      8730
Name: speeding, dtype: int64

In [11]:
df_collision.X.isna().value_counts()

False    191019
True       5097
Name: X, dtype: int64

In [12]:
df_collision.Y.isna().value_counts()

False    191019
True       5097
Name: Y, dtype: int64

In [13]:
df_collision.location.isna().value_counts()

False    191810
True       4306
Name: location, dtype: int64

In [14]:
df_collision.weather.isna().value_counts()

False    173400
True      22716
Name: weather, dtype: int64

In [14]:
df_collision.loc[df_collision.X.isna() & (-df_collision.location.isna()), columns].shape

(791, 4)

###### Conclusion: 
1. drop "speeding" variable
2. drop rows where "X" or "Y" is null, impute (X, Y) value by "location" value for 791 rows with Google's GeoCoding API?
3. impute "weather" value with other data sources

### Featur Engineering on Indicent Time Value

#### convert incident time field from string to datetime

In [15]:
df_collision['time'] = df_collision['incdttm'].apply(pd.to_datetime)

In [16]:
df_collision.head(2)

X          Y   objectid  inckey  coldetkey addrtype collisiontype  \
0 -122.342323  47.627572  162244138  233006     233406    Block       Head On   
1 -122.301639  47.630771  162244139  233085     233485    Block    Parked Car   

  exceptrsncode exceptrsndesc  fatalities         ...         st_colcode  \
0                         NaN           0         ...                 30   
1                         NaN           0         ...                 32   

                             st_coldesc   status  underinfl  vehcount  \
0  From Opposite Direction - All Others  Matched          N         2   
1               One Parked - One Moving  Matched          N         2   

                  weather seglanekey crosswalkkey  hitparkedcar  \
0  Clear or Partly Cloudy          0            0             N   
1                Overcast          0            0             Y   

                 time  
0 2016-11-04 09:55:00  
1 2016-12-01 10:33:00  

[2 rows x 41 columns]

#### Function to round incdttm field to its nearest hour value

source: https://stackoverflow.com/questions/48107644/rounding-datetime-to-the-nearest-hour?rq=1

In [17]:
def round_to_hour(dt):
    dt_start_of_hour = dt.replace(minute=0, second=0, microsecond=0)
    dt_half_hour = dt.replace(minute=30, second=0, microsecond=0)

    if dt >= dt_half_hour:
        # round up
        dt = dt_start_of_hour + datetime.timedelta(hours=1)
    else:
        # round down
        dt = dt_start_of_hour

    return dt

In [18]:
df_collision['hour'] = df_collision['time'].apply(round_to_hour)

In [19]:
columns = ['X', 'Y', 'addrtype', 'incdttm', 'lightcond', 'location', 'sdot_coldesc'
          , 'severitydesc', 'weather', 'time', 'hour']

df_collision.loc[:, columns].head(2)

X          Y addrtype                incdttm lightcond  \
0 -122.342323  47.627572    Block   11/4/2016 9:55:00 AM  Daylight   
1 -122.301639  47.630771    Block  12/1/2016 10:33:00 AM  Daylight   

                                        location  \
0      DEXTER AVE N BETWEEN ALOHA ST AND WARD ST   
1  24TH AVE E BETWEEN E HIGHLAND DR AND E LEE ST   

                                        sdot_coldesc  \
0  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
1  MOTOR VEHICLE STRUCK MOTOR VEHICLE, LEFT SIDE ...   

                     severitydesc                 weather                time  \
0  Property Damage Only Collision  Clear or Partly Cloudy 2016-11-04 09:55:00   
1  Property Damage Only Collision                Overcast 2016-12-01 10:33:00   

                 hour  
0 2016-11-04 10:00:00  
1 2016-12-01 11:00:00

#### convert seattle time to epoch time

In [20]:
from pytz import timezone

In [21]:
seattle_timezone = timezone('US/Pacific')

In [22]:
def convert_pacific_time_to_epoch(dt):
    return int(seattle_timezone.localize(dt).timestamp())

In [23]:
df_collision['epoch'] = df_collision['hour'].apply(convert_pacific_time_to_epoch)

### Save the resulting pandas dataframe to a SQL database

In [24]:
import sqlite3
from pandas.io import sql
# Create your connection.
conn = sqlite3.connect('collisions.db')

save the dataframe into sqlite3 database

In [27]:
df_collision.to_sql('collision_data', conn)

drop table if it exists

In [26]:
c = conn.cursor()
c.execute('''DROP TABLE collision_data;''')

when reading in the dataframe from sqlite3 db, the datetime fields become strings

In [28]:
df = pd.read_sql_query("select * from collision_data limit 1;", conn)

### Retrieve Hourly Weather data with DarkSky API

In [82]:
mask = (-df_collision.X.isna()) & (-df_collision.Y.isna())

In [85]:
unique_location_hour = df_collision.loc[mask, ['X', 'Y', 'epoch']].drop_duplicates()

In [132]:
a = df_collision.loc[mask].groupby(['X', 'Y', 'epoch']).size()

In [135]:
a.value_counts()

1    189768
2       582
3        22
4         4
5         1
dtype: int64

In [87]:
API_KEY = '367695a538073f1d5494a95afe404aa8'

In [88]:
COLUMNS = ['Latitude', 'Longitude', 'EpochTime', 'Summary', 'DegreesFahrenheit', 'Humidity', 'Visibility',
           'WindBearing', 'WindGust', 'WindSpeed', "PrecipitationIntensity", "PrecipitationType"]

def get_weather_json_darksky(latitude, longitude, epoch_time):
    """ Calls the DarkSky API with the given latitude, longitude and epoch time to retrieve historical weather data.
    :param latitude: Float value representing latitude
    :param longitude: Float value representing longitude
    :param epoch_time: Numeric value representing
    :return: A darksky.forecast.Forecast object
    """
    response = forecast(API_KEY, latitude, longitude, int(epoch_time))
    return response

def get_weather_series_from_response(latitude, longitude, epoch_time, response):
    """ """
    response_keys = response._data.keys()

    row_data = [
                latitude,
                longitude,
                epoch_time,
                response['summary'] if 'summary' in response_keys else None,
                response['temperature'] if 'temperature' in response_keys else None,
                response['humidity'] if 'humidity' in response_keys else None,
                response['visibility'] if 'visibility' in response_keys else None,
                response['windBearing'] if 'windBearing' in response_keys else None,
                response['windGust'] if 'windGust' in response_keys else None,
                response['windSpeed'] if 'windSpeed' in response_keys else None,
                response['precipIntensity'] if 'precipIntensity' in response_keys else None,
                response['precipType'] if 'precipType' in response_keys else None
    ]
    row = pd.Series(row_data, index = COLUMNS)
    return row

In [124]:
# df_weather = pd.DataFrame(columns = COLUMNS)

# for i in range(unique_location_hour.shape[0]):
    
for i in range(695, unique_location_hour.shape[0]):
    response = get_weather_json_darksky(
                    unique_location_hour.iloc[i,:].Y
                    , unique_location_hour.iloc[i,:].X
                    , int(unique_location_hour.iloc[i,:].epoch)
                    ).currently
    series = get_weather_series_from_response(
                    unique_location_hour.iloc[i,:].Y
                    , unique_location_hour.iloc[i,:].X
                    , int(unique_location_hour.iloc[i,:].epoch)    
                    , response
                    )
    df_weather = df_weather.append(series, ignore_index=True)

KeyboardInterrupt: 

In [129]:
unique_location_hour.shape[0]

190377

In [127]:
i

42501

In [125]:
df_weather.shape

(42501, 12)

In [128]:
df_weather.tail()

Latitude   Longitude   EpochTime                Summary  \
42496  47.559560 -122.266928  1139212800                  Clear   
42497  47.670918 -122.388863  1139212800                  Clear   
42498  47.532214 -122.392690  1139385600               Overcast   
42499  47.722321 -122.354184  1136793600  Light Rain and Breezy   
42500  47.681658 -122.327489  1139598000                  Clear   

       DegreesFahrenheit  Humidity  Visibility WindBearing WindGust WindSpeed  \
42496              39.91      0.86        9.71         334     None      8.24   
42497              40.94      0.84        9.50         345     None     10.57   
42498              48.16      0.76        9.61         177     None      7.57   
42499              46.37      0.85        8.66         167     None     16.26   
42500              49.67      0.39        9.83          60     None      7.25   

      PrecipitationIntensity PrecipitationType  
42496                      0              None  
42497                      0              None  
42498                      0              None  
42499                 0.0301              rain  
42500                      0              None

In [77]:
response = get_weather_json_darksky(
                unique_location_hour.loc[i,:].Y
                , unique_location_hour.loc[i,:].X
                , int(unique_location_hour.loc[i,:].epoch)
                ).currently

response_keys = response._data.keys()

row_data = [
            response['summary'] if 'summary' in response_keys else None,
            response['temperature'] if 'temperature' in response_keys else None,
            response['humidity'] if 'humidity' in response_keys else None,
            response['visibility'] if 'visibility' in response_keys else None,
            response['windBearing'] if 'windBearing' in response_keys else None,
            response['windGust'] if 'windGust' in response_keys else None,
            response['windSpeed'] if 'windSpeed' in response_keys else None,
            response['precipIntensity'] if 'precipIntensity' in response_keys else None,
            response['precipType'] if 'precipType' in response_keys else None
]

response_keys

HTTPError: Bad response

In [79]:
unique_location_hour

X          Y       epoch
0      -122.342323  47.627572  1478278800
1      -122.301639  47.630771  1480618800
2      -122.290784  47.719253  1482526800
3      -122.322535  47.623135  1482364800
4      -122.376803  47.689665  1482505200
5      -122.299223  47.618915  1480784400
6      -122.392486  47.580801  1482307200
7      -122.309670  47.595154  1482400800
8      -122.317791  47.657942  1483228800
9      -122.333141  47.662629  1482555600
10     -122.280805  47.525184  1482476400
11     -122.316764  47.622539  1481356800
12     -122.348641  47.696874  1478188800
13     -122.340185  47.731487  1480716000
14     -122.317448  47.606205  1480762800
15     -122.325841  47.703160  1482613200
16     -122.344559  47.661385  1482566400
17     -122.318438  47.597505  1482051600
18     -122.356271  47.571375  1481486400
19     -122.295618  47.641164  1482462000
20     -122.372173  47.667888  1482480000
21     -122.306849  47.691236  1482627600
22             NaN        NaN  1479063600
23     -122.323484  47.734053  1479171600
24     -122.303620  47.620213  1481342400
25     -122.258419  47.502992  1482624000
26     -122.329737  47.692339  1479060000
27     -122.338498  47.665774  1482937200
28     -122.286708  47.658424  1482710400
29     -122.314224  47.614122  1482537600
...            ...        ...         ...
196086 -122.337745  47.614832  1520409600
196087 -122.347036  47.611726  1520924400
196088 -122.370769  47.661469  1521010800
196089 -122.320060  47.678768  1519977600
196090 -122.328592  47.706812  1519891200
196091 -122.355453  47.705084  1521702000
196092 -122.300906  47.660510  1521702000
196093 -122.307191  47.664860  1521874800
196094 -122.336545  47.699594  1521702000
196095 -122.372169  47.670169  1521270000
196096 -122.360480  47.676869  1521529200
196097 -122.324025  47.681980  1519632000
196098 -122.295636  47.667861  1520668800
196099 -122.330778  47.600434  1519804800
196100 -122.350072  47.614307  1520323200
196101 -122.342575  47.652707  1520409600
196102 -122.271503  47.519048  1519718400
196103 -122.376189  47.564559  1520755200
196104 -122.376597  47.536482  1521010800
196105 -122.298900  47.605233  1521010800
196106 -122.308317  47.599243  1521010800
196107 -122.317862  47.709174  1521010800
196108 -122.321439  47.547503  1520064000
196109 -122.346238  47.628030  1520236800
196110 -122.329679  47.599198  1519977600
196111 -122.377704  47.564078  1520236800
196112 -122.329077  47.571421  1520150400
196113 -122.320750  47.606213  1519545600
196114 -122.296764  47.602656  1519459200
196115 -122.317404  47.664937  1519977600

[194668 rows x 3 columns]

In [78]:
unique_location_hour.loc[i,:]

X                 NaN
Y                 NaN
epoch    1.479064e+09
Name: 22, dtype: float64

In [75]:
response_keys

dict_keys(['time', 'summary', 'icon', 'precipIntensity', 'precipProbability', 'precipType', 'temperature', 'apparentTemperature', 'dewPoint', 'humidity', 'pressure', 'windSpeed', 'windBearing', 'visibility'])

In [64]:
df_weather.shape

(22, 12)

In [36]:
df_weather.shape

(1, 12)